sale_invoice_policy/models/sale_order_line.py

```diff
@@ -1,68 +1,99 @@
-
     _inherit = "sale.order.line"
 
     @api.depends(
         "qty_invoiced",
         "qty_delivered",
         "product_uom_qty",
-        "order_id.state",
+        "state",
         "order_id.invoice_policy",
     )
     def _compute_qty_to_invoice(self):
-        invoice_policies = set(self.mapped("order_id.invoice_policy"))
-        line_by_id = {line.id: line for line in self}
-        done_lines = self.env["sale.order.line"].browse()
-        for invoice_policy in invoice_policies:
-            so_lines = (
-                self.with_context(invoice_policy=invoice_policy)
-                .filtered(lambda x, p=invoice_policy: x.order_id.invoice_policy == p)
-                .with_prefetch()
-            )
-            if so_lines:
-                done_lines |= so_lines
-                super(SaleOrderLine, so_lines)._compute_qty_to_invoice()
-                for line in so_lines:
-                    # due to the change of context in compute methods,
-                    # assign the value in the modified context to self
-                    line_by_id[line.id].qty_to_invoice = line.qty_to_invoice
-        # Not to break function if (it could not happen) some records
-        # were not in so_lines
-        super(SaleOrderLine, self - done_lines)._compute_qty_to_invoice()
+        other_lines = self.filtered(
+            lambda l: l.product_id.type == "service" or not l.order_id.invoice_policy
+        )
+        super(SaleOrderLine, other_lines)._compute_qty_to_invoice()
+        for line in self - other_lines:
+            invoice_policy = line.order_id.invoice_policy
+            if invoice_policy == "order":
+                line.qty_to_invoice = line.product_uom_qty - line.qty_invoiced
+            else:
+                line.qty_to_invoice = line.qty_delivered - line.qty_invoiced
         return True
 
     @api.depends(
         "state",
-        "product_uom_qty",
+        "price_reduce",
+        "product_id",
+        "untaxed_amount_invoiced",
         "qty_delivered",
-        "qty_to_invoice",
-        "qty_invoiced",
+        "product_uom_qty",
         "order_id.invoice_policy",
     )
-    def _compute_invoice_status(self):
-        invoice_policies = set(self.mapped("order_id.invoice_policy"))
-        line_by_id = {line.id: line for line in self}
-        done_lines = self.env["sale.order.line"].browse()
-        for invoice_policy in invoice_policies:
-            so_lines = (
-                self.with_context(invoice_policy=invoice_policy)
-                .filtered(lambda x, p=invoice_policy: x.order_id.invoice_policy == p)
-                .with_prefetch()
+    def _compute_untaxed_amount_to_invoice(self):
+        other_lines = self.filtered(
+            lambda line: line.product_id.type == "service"
+            or not line.order_id.invoice_policy
+            or line.order_id.invoice_policy == line.product_id.invoice_policy
+            or line.state not in ["sale", "done"]
```

Comment:
Same here, I think you need to also check if `l.order_id.invoice_policy_required` is set on `False`, because if it is, you don't need to modify the line, since it is not required.